hypervisor/boot/acpi.c

```diff
@@ -140,24 +141,17 @@ static void *get_rsdp(void)
 
 #ifdef CONFIG_EFI_STUB
 	rsdp = get_rsdp_from_uefi();
-	if (rsdp) {
-		return rsdp;
-	}
 #endif
-
-	/* EBDA is addressed by the 16 bit pointer at 0x40E */
-	addr = (uint16_t *)hpa2hva(0x40eUL);
-
-	rsdp = biosacpi_search_rsdp((char *)hpa2hva((uint64_t)(*addr) << 4U),
-									0x400);
-	if (rsdp != NULL) {
-		return rsdp;
-	}
-
-	/* Check the upper memory BIOS space, 0xe0000 - 0xfffff. */
-	rsdp = biosacpi_search_rsdp((char *)hpa2hva(0xe0000UL), 0x20000);
-	if (rsdp != NULL) {
-		return rsdp;
+	if (rsdp == NULL) {
+		/* EBDA is addressed by the 16 bit pointer at 0x40E */
+		addr = (uint16_t *)hpa2hva(0x40eUL);
+
+		rsdp = biosacpi_search_rsdp((char *)hpa2hva((uint64_t)(*addr) << 4U),
+										0x400);
+		if (rsdp == NULL) {
+			/* Check the upper memory BIOS space, 0xe0000 - 0xfffff. */
+			rsdp = biosacpi_search_rsdp((char *)hpa2hva(0xe0000UL), 0x20000);
+		}
```

Comment:
@dongyaozu sorry, I didn't get it. Could you please show the details?

Reply:
if CONFIG_EFI_STUB is not true, the original code doesn't have if (rsdp==NULL) check before addr = xxxx.

Reply:
in original code, the rsdp is initialized to NULL.
so if CONFIG_EFI_STUB is not true, the code will start from if (rsdp == NULL).
if CONFIG_EFI_STUB is true, the code will found rsdp from multiboot info first, if not found, then it will continue searching.
If keep original semantic, we have two choices:
1.
static void *get_rsdp(void)
{
	struct acpi_table_rsdp *rsdp = NULL;
	uint16_t *addr;

#ifdef CONFIG_EFI_STUB
	rsdp = get_rsdp_from_uefi();
	if (rsdp == NULL) {
#endif
		/* EBDA is addressed by the 16 bit pointer at 0x40E */
		addr = (uint16_t *)hpa2hva(0x40eUL);

		rsdp = found_rsdp((char *)hpa2hva((uint64_t)(*addr) << 4U), 0x400);
		if (rsdp == NULL) {
			/* Check the upper memory BIOS space, 0xe0000 - 0xfffff. */
			rsdp = found_rsdp((char *)hpa2hva(0xe0000UL), 0x20000);
		}
#ifdef CONFIG_EFI_STUB
	}
#endif
	return rsdp;
}
The code would have a indent issue.

2. 
static void *get_rsdp(void)
{
	struct acpi_table_rsdp *rsdp = NULL;
	uint16_t *addr;

#ifdef CONFIG_EFI_STUB
	rsdp = get_rsdp_from_uefi();
	if (rsdp == NULL) {
		/* EBDA is addressed by the 16 bit pointer at 0x40E */
		addr = (uint16_t *)hpa2hva(0x40eUL);

		rsdp = found_rsdp((char *)hpa2hva((uint64_t)(*addr) << 4U), 0x400);
		if (rsdp == NULL) {
			/* Check the upper memory BIOS space, 0xe0000 - 0xfffff. */
			rsdp = found_rsdp((char *)hpa2hva(0xe0000UL), 0x20000);
		}
	}
#else
	/* EBDA is addressed by the 16 bit pointer at 0x40E */
	addr = (uint16_t *)hpa2hva(0x40eUL);

	rsdp = found_rsdp((char *)hpa2hva((uint64_t)(*addr) << 4U), 0x400);
	if (rsdp == NULL) {
		/* Check the upper memory BIOS space, 0xe0000 - 0xfffff. */
		rsdp = found_rsdp((char *)hpa2hva(0xe0000UL), 0x20000);
	}
#endif
	return rsdp;
}

What do you prefer?

Reply:
I see. So even the original code is fine.  Up to your choice

Reply:
Thanks @dongyaozu , I'd like to keep the original code. Please approve my forced update, thanks again.